SQLServerBulkCSVFileRecord drops empty trailing columns when escapeDelimiters = true

## Driver version
8.4.1

## SQL Server version
Microsoft SQL Server 2016 (SP2-CU8) (KB4505830) - 13.0.5426.0 (X64) 

## Client Operating System
Windows 10

## JAVA/JVM version
Amazon Corretto 11

## Table schema
```
CREATE TABLE [exclu].[oig_leie](
	[LASTNAME] [varchar](20),
	[FIRSTNAME] [varchar](15),
	[MIDNAME] [varchar](15),
	[BUSNAME] [varchar](30),
	[GENERAL] [varchar](20),
	[SPECIALTY] [varchar](20),
	[UPIN] [varchar](6),
	[NPI] [varchar](10),
	[DOB] [date],
	[ADDRESS] [varchar](30),
	[CITY] [varchar](20),
	[STATE] [varchar](2),
	[ZIP] [varchar](5),
	[EXCLTYPE] [varchar](9),
	[EXCLDATE] [date],
	[REINDATE] [varchar](8),
	[WAIVERDATE] [varchar](8),
	[WVRSTATE] [varchar](2)
)
```

## Problem description
1. Expected behaviour: CSV with empty last column would load successfully when `SQLServerBulkCSVFileRecord.setEscapeColumnDelimitersCSV(true)`.
2. Actual behaviour: Exception thrown because number of columns in data does not match number of columns in header
3. Error message/stack trace: 
```
Exception in thread "main" com.microsoft.sqlserver.jdbc.SQLServerException: Unable to retrieve data from the source.
	at com.microsoft.sqlserver.jdbc.SQLServerBulkCopy.writeBatchData(SQLServerBulkCopy.java:3626)
	at com.microsoft.sqlserver.jdbc.SQLServerBulkCopy.doInsertBulk(SQLServerBulkCopy.java:1566)
	at com.microsoft.sqlserver.jdbc.SQLServerBulkCopy.access$300(SQLServerBulkCopy.java:64)
	at com.microsoft.sqlserver.jdbc.SQLServerBulkCopy$1InsertBulk.doExecute(SQLServerBulkCopy.java:662)
	at com.microsoft.sqlserver.jdbc.TDSCommand.execute(IOBuffer.java:7375)
	at com.microsoft.sqlserver.jdbc.SQLServerConnection.executeCommand(SQLServerConnection.java:3206)
	at com.microsoft.sqlserver.jdbc.SQLServerBulkCopy.sendBulkLoadBCP(SQLServerBulkCopy.java:696)
	at com.microsoft.sqlserver.jdbc.SQLServerBulkCopy.writeToServer(SQLServerBulkCopy.java:1654)
	at com.microsoft.sqlserver.jdbc.SQLServerBulkCopy.writeToServer(SQLServerBulkCopy.java:619)
	... 2 more
Caused by: com.microsoft.sqlserver.jdbc.SQLServerException: Source data does not match source schema.
	at com.microsoft.sqlserver.jdbc.SQLServerBulkCSVFileRecord.getRowData(SQLServerBulkCSVFileRecord.java:267)
	at com.microsoft.sqlserver.jdbc.SQLServerBulkCopy.writeBatchData(SQLServerBulkCopy.java:3623)
	... 10 more
```
4. Any other details that can be helpful:
`-1` needs to be added to the `split` invocation at line 200 and 247 of SQLServerBulkCSVFileRecord.java to mirror the `split` invocation when `escapeDelimiters` is `false`, otherwise blank columns at the end of the line are dropped.  E.g., `escapeQuotesRFC4180(currentLine.split(delimiter + escapeSplitPattern, -1))`

## Reproduction code
```
URL leieCSV = new URL("https://oig.hhs.gov/exclusions/downloadables/UPDATED.csv");
URLConnection leieCSVConnection = leieCSV.openConnection();

try(SQLServerBulkCopy bulkCopy = new SQLServerBulkCopy(conn);
    SQLServerBulkCSVFileRecord fileRecord = new SQLServerBulkCSVFileRecord(leieCSVConnection.getInputStream(), leieCSVConnection.getContentEncoding(), ",", true))
{
    fileRecord.setEscapeColumnDelimitersCSV(true);
    fileRecord.addColumnMetadata(1, "LASTNAME", java.sql.Types.VARCHAR, 20, 0);
    fileRecord.addColumnMetadata(2, "FIRSTNAME", java.sql.Types.VARCHAR, 15, 0);
    fileRecord.addColumnMetadata(3, "MIDNAME", java.sql.Types.VARCHAR, 15, 0);
    fileRecord.addColumnMetadata(4, "BUSNAME", java.sql.Types.VARCHAR, 30, 0);
    fileRecord.addColumnMetadata(5, "GENERAL", java.sql.Types.VARCHAR, 20, 0);
    fileRecord.addColumnMetadata(6, "SPECIALTY", java.sql.Types.VARCHAR, 20, 0);
    fileRecord.addColumnMetadata(7, "UPIN", java.sql.Types.VARCHAR, 6, 0);
    fileRecord.addColumnMetadata(8, "NPI", java.sql.Types.VARCHAR, 10, 0);
    fileRecord.addColumnMetadata(9, "DOB", java.sql.Types.DATE, 0, 0);
    fileRecord.addColumnMetadata(10, "ADDRESS", java.sql.Types.VARCHAR, 30, 0);
    fileRecord.addColumnMetadata(11, "CITY", java.sql.Types.VARCHAR, 20, 0);
    fileRecord.addColumnMetadata(12, "STATE", java.sql.Types.VARCHAR, 2, 0);
    fileRecord.addColumnMetadata(13, "ZIP", java.sql.Types.VARCHAR, 5, 0);
    fileRecord.addColumnMetadata(14, "EXCLTYPE", java.sql.Types.VARCHAR, 9, 0);
    fileRecord.addColumnMetadata(15, "EXCLDATE", java.sql.Types.DATE, 0, 0);
    fileRecord.addColumnMetadata(16, "REINDATE", java.sql.Types.VARCHAR, 8, 0);
    fileRecord.addColumnMetadata(17, "WAIVERDATE", java.sql.Types.VARCHAR, 8, 0);
    fileRecord.addColumnMetadata(18, "WVRSTATE", java.sql.Types.VARCHAR, 2, 0);
    
    stmt.executeUpdate("TRUNCATE TABLE exclu.oig_leie");
    bulkCopy.setDestinationTableName("exclu.oig_leie");
    bulkCopy.writeToServer(fileRecord);
}
```

--- a/src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkCSVFileRecord.java
+++ b/src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkCSVFileRecord.java
@@ -196,7 +196,7 @@ public class SQLServerBulkCSVFileRecord extends SQLServerBulkRecord implements j
             if (null != currentLine) {
                 columnNames = (escapeDelimiters && currentLine.contains("\""))
                                                                                ? escapeQuotesRFC4180(currentLine.split(
-                                                                                       delimiter + escapeSplitPattern))
+                                                                                       delimiter + escapeSplitPattern, -1))
                                                                                : currentLine.split(delimiter, -1);
             }
         }
@@ -243,7 +243,7 @@ public class SQLServerBulkCSVFileRecord extends SQLServerBulkRecord implements j
              */
             String[] data = (escapeDelimiters && currentLine.contains("\""))
                                                                              ? escapeQuotesRFC4180(currentLine.split(
-                                                                                     delimiter + escapeSplitPattern))
+                                                                                     delimiter + escapeSplitPattern, -1))
                                                                              : currentLine.split(delimiter, -1);
 
             // Cannot go directly from String[] to Object[] and expect it to act as an array.

--- a/src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkCSVFileRecord.java
+++ b/src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkCSVFileRecord.java
@@ -194,10 +194,8 @@ public class SQLServerBulkCSVFileRecord extends SQLServerBulkRecord implements j
         if (firstLineIsColumnNames) {
             currentLine = fileReader.readLine();
             if (null != currentLine) {
-                columnNames = (escapeDelimiters && currentLine.contains("\""))
-                                                                               ? escapeQuotesRFC4180(currentLine.split(
-                                                                                       delimiter + escapeSplitPattern, -1))
-                                                                               : currentLine.split(delimiter, -1);
+                columnNames = (escapeDelimiters && currentLine.contains("\"")) ? escapeQuotesRFC4180(
+                        currentLine.split(delimiter + escapeSplitPattern, -1)) : currentLine.split(delimiter, -1);
             }
         }
     }
@@ -241,10 +239,8 @@ public class SQLServerBulkCSVFileRecord extends SQLServerBulkRecord implements j
              * Binary data may be corrupted The limit in split() function should be a negative value, otherwise trailing
              * empty strings are discarded. Empty string is returned if there is no value.
              */
-            String[] data = (escapeDelimiters && currentLine.contains("\""))
-                                                                             ? escapeQuotesRFC4180(currentLine.split(
-                                                                                     delimiter + escapeSplitPattern, -1))
-                                                                             : currentLine.split(delimiter, -1);
+            String[] data = (escapeDelimiters && currentLine.contains("\"")) ? escapeQuotesRFC4180(
+                    currentLine.split(delimiter + escapeSplitPattern, -1)) : currentLine.split(delimiter, -1);
 
             // Cannot go directly from String[] to Object[] and expect it to act as an array.
 

--- a/src/test/java/com/microsoft/sqlserver/jdbc/bulkCopy/BulkCopyCSVTest.java
+++ b/src/test/java/com/microsoft/sqlserver/jdbc/bulkCopy/BulkCopyCSVTest.java
@@ -149,21 +149,21 @@ public class BulkCopyCSVTest extends AbstractTest {
         String tableName = AbstractSQLGenerator.escapeIdentifier(RandomUtil.getIdentifier("BulkEscape"));
         String fileName = filePath + inputFileDelimiterEscape;
         /*
-         * The list below is the copy of inputFileDelimiterEscape with quotes removed.
+         * The list below is the copy of inputFileDelimiterEsc ape with quotes removed.
          */
         String[][] expectedEscaped = new String[11][4];
-        expectedEscaped[0] = new String[] {"test", " test\"", "no@split", " testNoQuote"};
-        expectedEscaped[1] = new String[] {null, null, null, null};
-        expectedEscaped[2] = new String[] {"\"", "test\"test", "test@\"  test", null};
+        expectedEscaped[0] = new String[] {"test", " test\"", "no@split", " testNoQuote", ""};
+        expectedEscaped[1] = new String[] {null, null, null, null, ""};
+        expectedEscaped[2] = new String[] {"\"", "test\"test", "test@\"  test", null, ""};
         expectedEscaped[3] = new String[] {"testNoQuote  ", " testSpaceAround ", " testSpaceInside ",
-                "  testSpaceQuote\" "};
-        expectedEscaped[4] = new String[] {null, null, null, " testSpaceInside "};
-        expectedEscaped[5] = new String[] {"1997", "Ford", "E350", "E63"};
-        expectedEscaped[6] = new String[] {"1997", "Ford", "E350", "E63"};
-        expectedEscaped[7] = new String[] {"1997", "Ford", "E350", "Super@ luxurious truck"};
-        expectedEscaped[8] = new String[] {"1997", "Ford", "E350", "Super@ \"luxurious\" truck"};
-        expectedEscaped[9] = new String[] {"1997", "Ford", "E350", "E63"};
-        expectedEscaped[10] = new String[] {"1997", "Ford", "E350", " Super luxurious truck "};
+                "  testSpaceQuote\" ", ""};
+        expectedEscaped[4] = new String[] {null, null, null, " testSpaceInside ", ""};
+        expectedEscaped[5] = new String[] {"1997", "Ford", "E350", "E63", ""};
+        expectedEscaped[6] = new String[] {"1997", "Ford", "E350", "E63", ""};
+        expectedEscaped[7] = new String[] {"1997", "Ford", "E350", "Super@ luxurious truck", ""};
+        expectedEscaped[8] = new String[] {"1997", "Ford", "E350", "Super@ \"luxurious\" truck", ""};
+        expectedEscaped[9] = new String[] {"1997", "Ford", "E350", "E63", ""};
+        expectedEscaped[10] = new String[] {"1997", "Ford", "E350", " Super luxurious truck ", ""};
 
         try (Connection con = getConnection(); Statement stmt = con.createStatement();
                 SQLServerBulkCopy bulkCopy = new SQLServerBulkCopy(con);
@@ -176,8 +176,9 @@ public class BulkCopyCSVTest extends AbstractTest {
             fileRecord.addColumnMetadata(3, null, java.sql.Types.VARCHAR, 50, 0);
             fileRecord.addColumnMetadata(4, null, java.sql.Types.VARCHAR, 50, 0);
             fileRecord.addColumnMetadata(5, null, java.sql.Types.VARCHAR, 50, 0);
+            fileRecord.addColumnMetadata(6, null, java.sql.Types.VARCHAR, 50, 0);
             stmt.executeUpdate("CREATE TABLE " + tableName
-                    + " (id INT IDENTITY(1,1), c1 VARCHAR(50), c2 VARCHAR(50), c3 VARCHAR(50), c4 VARCHAR(50))");
+                    + " (id INT IDENTITY(1,1), c1 VARCHAR(50), c2 VARCHAR(50), c3 VARCHAR(50), c4 VARCHAR(50), c5 VARCHAR(50))");
             bulkCopy.writeToServer(fileRecord);
 
             int i = 0;
@@ -193,10 +194,10 @@ public class BulkCopyCSVTest extends AbstractTest {
             }
         }
     }
-    
+
     /**
-     * test simple csv file for bulkcopy, for GitHub issue 1391
-     * Tests to ensure that the set returned by getColumnOrdinals doesn't have to be ordered
+     * test simple csv file for bulkcopy, for GitHub issue 1391 Tests to ensure that the set returned by
+     * getColumnOrdinals doesn't have to be ordered
      */
     @Test
     @DisplayName("Test SQLServerBulkCSVFileRecord GitHb 1391")
@@ -212,7 +213,7 @@ public class BulkCopyCSVTest extends AbstractTest {
             fail(e.getMessage());
         }
     }
-    
+
     // Used for testing issue reported in https://github.com/microsoft/mssql-jdbc/issues/1391
     private class BulkData1391 extends SQLServerBulkCSVFileRecord {
 

--- a/src/test/resources/BulkCopyCSVTestInputDelimiterEscape.csv
+++ b/src/test/resources/BulkCopyCSVTestInputDelimiterEscape.csv
@@ -1,11 +1,11 @@
-1@"test"@ " test"""@ "no@split" @ testNoQuote
-2@""@  ""@ ""@ ""
-3@""""@ "test""test"@ "test@""  test"@ ""
-4@testNoQuote  @ testSpaceAround @ " testSpaceInside "@  "  testSpaceQuote"" "
-5@""@  ""@ ""@ " testSpaceInside "
-6@1997@Ford@E350@E63
-7@"1997"@"Ford"@"E350"@"E63"
-8@1997@Ford@E350@"Super@ luxurious truck"
-9@1997@Ford@E350@"Super@ ""luxurious"" truck"
-10@1997@ "Ford" @E350@  "E63"
-11@1997@Ford@E350@" Super luxurious truck "
+1@"test"@ " test"""@ "no@split" @ testNoQuote@ 
+2@""@  ""@ ""@ ""@ 
+3@""""@ "test""test"@ "test@""  test"@ ""@ 
+4@testNoQuote  @ testSpaceAround @ " testSpaceInside "@  "  testSpaceQuote"" "@ 
+5@""@  ""@ ""@ " testSpaceInside "@ 
+6@1997@Ford@E350@E63@ 
+7@"1997"@"Ford"@"E350"@"E63"@ 
+8@1997@Ford@E350@"Super@ luxurious truck"@ 
+9@1997@Ford@E350@"Super@ ""luxurious"" truck"@ 
+10@1997@ "Ford" @E350@  "E63"@ 
+11@1997@Ford@E350@" Super luxurious truck "@ 

